src/socket/tcp/congestion.rs

```diff
@@ -0,0 +1,93 @@
+use crate::time::Instant;
+
+use super::RttEstimator;
+
+#[cfg(feature = "cubic")]
+pub(super) mod cubic;
+
+pub(super) mod reno;
+
+#[allow(unused_variables)]
+pub(super) trait CongestionController {
+    /// Returns the number of bytes that can be sent.
+    fn window(&self) -> usize;
+
+    /// Set the remote window size.
+    fn set_remote_window(&mut self, remote_window: usize) {}
+
+    fn on_ack(&mut self, now: Instant, len: usize, rtt: &RttEstimator) {}
+
+    fn on_retransmit(&mut self, now: Instant) {}
+
+    fn on_duplicate_ack(&mut self, now: Instant) {}
+
+    fn pre_transmit(&mut self, now: Instant) {}
+
+    fn post_transmit(&mut self, now: Instant, len: usize) {}
+
+    /// Set the maximum segment size.
+    fn set_mss(&mut self, mss: usize) {}
+}
+
+#[derive(Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub(super) enum GenericController {
+    None,
+    Reno(reno::Reno),
+
+    #[cfg(feature = "cubic")]
+    Cubic(cubic::Cubic),
+}
+
+impl GenericController {
+    pub fn new() -> Self {
+        GenericController::None
+    }
+
+    fn inner_mut(&mut self) -> Option<&mut dyn CongestionController> {
+        match self {
+            GenericController::None => None,
+            GenericController::Reno(r) => Some(r),
+
+            #[cfg(feature = "cubic")]
+            GenericController::Cubic(c) => Some(c),
+        }
+    }
+
+    fn inner(&self) -> Option<&dyn CongestionController> {
+        match self {
+            GenericController::None => None,
+            GenericController::Reno(r) => Some(r),
+
+            #[cfg(feature = "cubic")]
+            GenericController::Cubic(c) => Some(c),
+        }
+    }
+}
+
+impl CongestionController for GenericController {
```

Comment:
So I was actually thinking of something else: where the `None` controller option would belong to just another implementation of a congestion controller. The enum itself would not implement `CongestionController` at all, and the `&mut dyn CongestionController` would be used by `tcp::Socket` itself.

Reply:
Use `GenericController`'s `inner()` and `inner_mut()` methods to call `CongestionController`'s methods.